util/fipstools/acvp/acvptool/subprocess/kdf-components.go

```diff
@@ -55,134 +65,282 @@ type kdfCompTestGroupResponse struct {
+	if err != nil {
+		return err
+	}
+
+	sessionId, err := hex.DecodeString(test.SessionIdHex)
+	if err != nil {
+		return err
+	}
+
+	// NIST SP 800-135 says that the lengths of the IVs and encryption keys are determined by the supported block ciphers
+	// We can look at https://github.com/usnistgov/ACVP-Server/blob/master/gen-val/json-files/kdf-components-ssh-1.0/internalProjection.json
+	// to determine what these lengths are
+	ivLen := ivLenMap[group.Cipher]
+	encyrptionKeyLen := encyrptionKeyLenMap[group.Cipher]
+	integrityKeyLen := integrityKeyLenMap[group.Hash]
+
+	// Definitions for last argument of Transact() can be found in aws-lc/include/openssl/sshkdf.h
+	initialIvClient, err := m.Transact(method, 1,
+		secretVal, hashVal, sessionId,
+		uint32le(65), uint32le(ivLen))
+	if err != nil {
+		return err
+	}
+
+	initialIvServer, err := m.Transact(method, 1,
+		secretVal, hashVal, sessionId,
+		uint32le(66), uint32le(ivLen))
+	if err != nil {
+		return err
+	}
+
+	encryptionKeyClient, err := m.Transact(method, 1,
+		secretVal, hashVal, sessionId,
+		uint32le(67), uint32le(encyrptionKeyLen))
+	if err != nil {
+		return err
+	}
+
+	encryptionKeyServer, err := m.Transact(method, 1,
+		secretVal, hashVal, sessionId,
+		uint32le(68), uint32le(encyrptionKeyLen))
+	if err != nil {
+		return err
+	}
+
+	integrityKeyClient, err := m.Transact(method, 1,
+		secretVal, hashVal, sessionId,
+		uint32le(69), uint32le(integrityKeyLen))
+	if err != nil {
+		return err
+	}
+
+	integrityKeyServer, err := m.Transact(method, 1,
+		secretVal, hashVal, sessionId,
+		uint32le(70), uint32le(integrityKeyLen))
+	if err != nil {
+		return err
+	}
+
+	response.Tests = append(response.Tests, kdfCompTestResponse{
+		ID:                     test.ID,
+		InitialIvClientHex:     hex.EncodeToString(initialIvClient[0]),
+		InitialIvServerHex:     hex.EncodeToString(initialIvServer[0]),
+		EncryptionKeyClientHex: hex.EncodeToString(encryptionKeyClient[0]),
+		EncryptionKeyServerHex: hex.EncodeToString(encryptionKeyServer[0]),
+		IntegrityKeyClientHex:  hex.EncodeToString(integrityKeyClient[0]),
+		IntegritykeyServerHex:  hex.EncodeToString(integrityKeyServer[0]),
+	})
+
+	return nil
+}
+
+func (k *kdfComp) Process(vectorSet []byte, m Transactable) (interface{}, error) {
+	var parsed kdfCompVectorSet
+	if err := json.Unmarshal(vectorSet, &parsed); err != nil {
+		return nil, err
+	}
+
+	// See https://pages.nist.gov/ACVP/draft-celi-acvp-kdf-tls.html
+	var ret []kdfCompTestGroupResponse
+	for _, group := range parsed.Groups {
+		response := kdfCompTestGroupResponse{
+			ID: group.ID,
+		}
+
+		method, err := ProcessHeader(parsed.Mode, k, group)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, test := range group.Tests {
+			if parsed.Mode == "ssh" {
+				err = HandleSSH(test, k, m, method, group, &response)
+			} else if parsed.Mode == "tls" || parsed.Mode == "KDF" {
+				err = HandleTLS(test, k, m, method, group, &response)
 			}
```

Comment:
Does this need to handle the case mode != ssh or tls?

Reply:
So as is if the script gets something where the case mode isn't described, it'll just output an empty json, which should fail any test. But it'd probably be a good idea to handle it more elegantly so I'll do that 👍